lib/active_merchant/billing/gateways/ixopay.rb

```diff
@@ -67,55 +54,193 @@ def scrub(transcript)
 
       private
 
-      def add_customer_data(post, options)
+      def headers(xml)
+        timestamp = Time.now.httpdate
+        signature = generate_signature('POST', xml, timestamp)
+
+        {
+          'Authorization' => "Gateway #{options[:api_key]}:#{signature}",
+          'Date' => timestamp,
+          'Content-Type' => 'text/xml; charset=utf-8'
+        }
+      end
+
+      def generate_signature(http_method, xml, timestamp)
+        content_type = 'text/xml; charset=utf-8'
+
+        message = http_method + "\n" + Digest::MD5.hexdigest(xml) + "\n" + content_type + "\n" \
+          + timestamp + "\n\n" + '/transaction'
+
+        digest = OpenSSL::Digest.new('sha512')
+        hmac = OpenSSL::HMAC.digest(digest, @secret, message)
+
+        Base64.encode64(hmac).split.join
+      end
+
+      def parse(action, xml)
+        parse_element({:action => action}, REXML::Document.new(xml))
+      end
+
+      # This generic method appears in a number of gateways that parse XML.
+      # In the future, we should investigate finding a library method to
+      # drop in, or factoring it out to a module or base class.
+      def parse_element(raw, node)
+        node_name = node.name.underscore
+
+        node.attributes.each do |k, v|
+          raw["#{node_name}_#{k.underscore}".to_sym] = v
+        end
+
+        if node.has_elements?
+          raw[node_name.to_sym] = true unless node.name.blank?
+          node.elements.each { |e| parse_element(raw, e) }
+        elsif node.children.count > 1
+          raw[node_name.to_sym] = node.children.join(' ').strip
+        else
+          raw[node_name.to_sym] = node.text unless node.text.nil?
+        end
+
+        raw
+      end
+
+      def build_purchase_request(money, payment_method, options)
+        xml = Builder::XmlMarkup.new(indent: 2)
+
+        xml.instruct! :xml, encoding: 'utf-8'
+
+        xml.tag! 'transactionWithCard', 'xmlns' => 'http://secure.ixopay.com/Schema/V2/TransactionWithCard' do
```

Comment:
Rather than creating a `do...end` block here, it seems to make more sense in the context to use `{...}`. (Similar to examples [in the API docs](https://apidock.com/rails/Builder/XmlMarkup))


You could also consider refactoring this to use the block structure:
```
xml.transactionWithCard { |xml| 
  xml.username(@options[:username]
  xml.password(Digest::SHA1.hexdigest(@options[:password])
  add_card_data(xml, payment_method)
  add_debit(xml, money, options)
}
```

Reply:
To answer the question of whether this something that is done elsewhere in Active Merchant, I found several examples across gateways:

- [Authorize.net](https://github.com/activemerchant/active_merchant/blob/2821fbdd5941af0880716aa43e23a18f3f26945e/lib/active_merchant/billing/gateways/authorize_net.rb#L166)
- [S5](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/s5.rb#L130) (I'm a big fan of this example: the code looks super slick, and it seems to have been added by Spreedly.)
- [Optimal Payment](https://github.com/activemerchant/active_merchant/blob/35e2bd338c2145ba3b6d26c3e4e473e23a5dc019/lib/active_merchant/billing/gateways/optimal_payment.rb#L210)
- [IPP](https://github.com/activemerchant/active_merchant/blob/776f0735d81d69ec46d943644dd65b175ba7ea10/lib/active_merchant/billing/gateways/ipp.rb#L101)
- [Bambora Apac](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/s5.rb#L112)

Based on these examples, the above code would more accurately be written:
```
xml.transactionWithCard do
  xml.username    @options[:username]
  xml.password    Digest::SHA1.hexdigest(@options[:password]
  add_card_data(xml, payment_method)
  add_debit(xml, money, options)
end
```

Reply:
Done.